src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,184 @@
+     * @param source the {@link AbstractGitSCMSource}
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public GitRepoSizeEstimator(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        boolean useAPI = false;
+
+        useCache = setSizeFromCache(source);
+
+        if (useCache) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            useAPI = setSizeFromAPI(source.getRemote());
+        }
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        }
+
+        if (!useAPI && !useCache) {
+            implementation = "DEFAULT";
+        }
+        determineGitTool(implementation);
+    }
+
+    /**
+     * Estimate size of a repository using the extension point
+     * @param remoteName: The URL of the repository
+     */
+    public GitRepoSizeEstimator(String remoteName) {
+        boolean useAPI = setSizeFromAPI(remoteName);
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            implementation = "DEFAULT";
+        }
+        determineGitTool(implementation);
+    }
+
+    /**
+     * Determine and estimate the size of a .git cached directory
+     * @param source: Use a {@link AbstractGitSCMSource} to access a cached Jenkins directory, we do not lock it.
+     * @return useCache
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private boolean setSizeFromCache(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        String cacheEntry = source.getCacheEntry();
+        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry);
+        if (cacheDir != null) {
+            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+            GitClient client = git.getClient();
+            if (!client.hasGitRepo()) {
+                useCache = false;
+            } else {
+                useCache = true;
+                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
+                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+            }
+        } else {
+            useCache = false;
+        }
+        return useCache;
+    }
+
+    /**
+     * Check if the desired implementation of extension is present and ask for the size of repository if it does
+     * @param repoUrl: The remote name derived from {@link GitSCMSource} object
+     * @return boolean useAPI or not.
+     */
+    private boolean setSizeFromAPI(String repoUrl) {
+        List<RepositorySizeAPI> acceptedRepository = Objects.requireNonNull(RepositorySizeAPI.all())
+                .stream()
+                .filter(r -> r.acceptsRemote(repoUrl))
+                .collect(Collectors.toList());
```

Comment:
Maybe, this filter might be done in a method of the Extension, so here we have something like
```suggestion
        List<RepositorySizeAPI> acceptedRepository = RepositorySizeAPI.getValids(repoUrl);
```